modeling/theoretical-simulator.go

```diff
@@ -27,48 +30,142 @@ func (sim TheoreticalSimulator) Simulate(region regionInfo, endpointSlices map[s
 	if len(region.zoneDetails) == 0 || len(endpointSlices) == 0 {
 		return SimulationResult{}, errors.New("can't evaluate probability based on empty zones or endpointslices")
 	}
-	// kube-proxy calculation, zones name - endpoints in potential destination sclies
-	kubeProxy := make(map[string]map[string]float64)
-	for name := range region.zoneDetails {
-		kubeProxy[name] = map[string]float64{}
-		for _, slice := range endpointSlices {
-			for compZone, pods := range slice.Composition {
-				kubeProxy[name][compZone] += float64(pods) * slice.Weights[name]
+	// weighted number of endpoints in each sliceGroup can be reached by every
+	// zone
+	// first key: zone name
+	// second key: sliceGroup label
+	zoneReachableEndpointsInSG := make(map[string]map[string]float64)
+	for zone := range region.zoneDetails {
+		zoneReachableEndpointsInSG[zone] = map[string]float64{}
+		for sliceLabel, slice := range endpointSlices {
+			zoneReachableEndpointsInSG[zone][sliceLabel] = float64(slice.numberOfEndpoints()) * slice.ZoneTrafficWeights[zone]
+		}
+	}
+	// total weighted number of endpoints one zone can send traffic to
+	zoneReachableEndpoints := map[string]float64{}
+	for zone, sliceGroup := range zoneReachableEndpointsInSG {
+		for label := range sliceGroup {
+			zoneReachableEndpoints[zone] += sliceGroup[label]
+		}
+	}
+	// ratio of traffic received by a sliceGroup sent by each zone
+	zoneTrafficToSG := map[string]map[string]float64{}
+	for zone := range region.zoneDetails {
+		zoneTrafficToSG[zone] = map[string]float64{}
+		for label := range endpointSlices {
+			zoneTrafficToSG[zone][label] = zoneReachableEndpointsInSG[zone][label] / zoneReachableEndpoints[zone]
+		}
+	}
+	// total ratio of traffic received by each EndpointSliceGroup
+	sgTrafficRatio := map[string]float64{}
+	for label := range endpointSlices {
+		for zone := range region.zoneDetails {
+			sgTrafficRatio[label] += region.zoneDetails[zone].nodesRatio * zoneTrafficToSG[zone][label]
+		}
+	}
+	// endpoints traffic load
+	// first key: zone the endpoints belong to
+	// second key: sliceGroup the endpoints assigned to
```

Comment:
I think this pattern is used 3x in this function - maybe we could use a struct that contained all this information for a sliceGroup instead of 3 separate maps?

Reply:
Done